packages/core/src/render3/after_render_hooks.ts

```diff
@@ -0,0 +1,212 @@
+  }
+
+  let destroy: VoidFunction|undefined;
+  const unregisterFn = inject(DestroyRef).onDestroy(() => destroy?.());
+  const manager = inject(AfterRenderHooksManager);
+  const hook = new AfterRenderHook(callback);
+
+  destroy = () => {
+    manager.unregister(hook);
+    unregisterFn();
+  };
+  manager.register(hook);
+  return {destroy};
+}
+
+/**
+ * Register a callback to be invoked the next time the application
+ * finishes rendering.
+ *
+ * Note that the callback will run
+ * - in the order it was registered
+ * - on browser platforms only
+ *
+ * @param callback A callback function to register
+ *
+ * @usageNotes
+ *
+ * Use `afterNextRender` to read or write the DOM once,
+ * for example to initialize a non-Angular library.
+ *
+ * ### Example
+ * ```ts
+ * @Component({
+ *   selector: 'my-chart-cmp',
+ *   template: `<div #chart>{{ ... }}</div>`,
+ * })
+ * export class MyChartCmp {
+ *   @ViewChild('chart') chartRef: ElementRef;
+ *   chart: MyChart|null;
+ *
+ *   constructor() {
+ *     afterNextRender(() => {
+ *       this.chart = new MyChart(this.chartRef.nativeElement);
+ *     });
+ *   }
+ * }
+ * ```
+ *
+ * @developerPreview
+ */
+export function afterNextRender(callback: VoidFunction): AfterRenderHookRef {
+  ngDevMode && assertInInjectionContext(afterNextRender);
+
+  if (!isPlatformBrowser()) {
+    return {destroy() {}};
+  }
+
+  let destroy: VoidFunction|undefined;
+  const unregisterFn = inject(DestroyRef).onDestroy(() => destroy?.());
+  const manager = inject(AfterRenderHooksManager);
+  const hookInvokedOnce = new AfterRenderHook(() => {
+    destroy?.();
+    callback();
+  });
+
+  destroy = () => {
+    manager.unregister(hookInvokedOnce);
+    unregisterFn();
+  };
+  manager.register(hookInvokedOnce);
+  return {destroy};
+}
+
+/**
+ * A wrapper around a function to be used as an after render hook.
+ */
+class AfterRenderHook {
+  private callback: VoidFunction;
+
+  constructor(callback: VoidFunction) {
+    this.callback = callback;
+  }
+
+  invoke() {
+    this.callback();
+  }
+}
+
+/**
+ * Implements `afterRender` callback manager logic.
+ */
+export class AfterRenderHooksManager {
+  private hooks: Set<AfterRenderHook> = new Set();
+  private renderDepth: number = 0;
+  private runningCallbacks: boolean = false;
+
+  /**
+   * Mark the beginning of a render operation (i.e. CD cycle).
+   * Throws if called from an `afterRender` callback.
+   */
+  begin() {
+    if (this.runningCallbacks) {
+      throw new RuntimeError(
+          RuntimeErrorCode.RECURSIVE_APPLICATION_RENDER,
+          'A render operation began before the previous operation ended.');
+    }
+
+    this.renderDepth++;
+  }
+
+  /**
+   * Mark the end of a render operation. Registered hooks
+   * are invoked if there are no more pending operations.
+   */
+  end() {
+    this.renderDepth--;
+
+    if (this.renderDepth === 0) {
+      try {
+        this.runningCallbacks = true;
+        this.hooks.forEach(hook => hook.invoke());
```

Comment:
nit: we can consider doing something like this to avoid a situation when running a hook results in modifying an array that is being iterated over:
```suggestion
        const hooks = this.hooks;
        // Reset hooks array to avoid mutations while invoking hooks.
        this.hooks = [];
        hooks.forEach(hook => hook.invoke());
```

Reply:
`hooks` is a `Set`, so it's fine to modify it during iteration. Do you have a specific case in mind where that shouldn't be allowed?

Reply:
I was thinking about edge cases when `afterRender` would be used inside another `afterRender`, for example:
```
afterRender(() => {
  afterRender(() => { ... });
});
```
I guess in this case the inner `afterRender` would add a callback fn to the `hooks` Set and the callback would be invoked right away (when an expected behavior would be to call it later, after a subsequent render).